src/meta/src/hummock/manager/mod.rs

```diff
@@ -196,6 +200,78 @@ where
         Ok(instance)
     }
 
+    pub async fn start_compaction_heartbeat(
+        hummock_manager: Arc<Self>,
+    ) -> (JoinHandle<()>, Sender<()>) {
+        let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
+        let compactor_manager = hummock_manager.compactor_manager.clone();
+        let join_handle = tokio::spawn(async move {
+            let mut min_interval = tokio::time::interval(std::time::Duration::from_millis(1000));
+            loop {
+                tokio::select! {
+                    // Wait for interval
+                    _ = min_interval.tick() => {
+                    },
+                    // Shutdown
+                    _ = &mut shutdown_rx => {
+                        tracing::info!("Compaction heartbeat checker is stopped");
+                        return;
+                    }
+                }
+                for (context_id, mut task) in compactor_manager.get_timed_out_tasks() {
+                    if let Some(compactor) = compactor_manager.get_compactor(context_id) {
+                        let _ = compactor.cancel_task(task.task_id).await;
+                    }
+                    // Since we cannot rely on the compactor to return, we will try to update the
+                    // compact task status ourselves which is an idempotent,
+                    // serializable operation.
+
+                    // Change task status to failed
+                    task.task_status = false;
+
+                    // TODO: this needs to be reliable, in other words, we cannot tolerate errors.
+                    // So, we retry and also have an idempotent fail-all which
+                    // we also retry (manually trigger `try_send_compaction_request`).
+                    // If all else fails, we report that we may never recover from this failure and
+                    // may fail to balance the LSM tree.
+                    let mut success = false;
+                    let mut last_err = None;
+                    for attempt_id in 0..REPORT_COMPACT_TASK_RETRIES {
```

Comment:
If we don't consume tasks in get_timed_out_tasks, then the retry here is optional and report_compact_task will succeed in later loop eventually.

Reply:
Hmm, we could indeed not consume if not successful. I think that is the better behaviour. And the retry behaviour occurs quite regularly, so perhaps we can get rid of all the retry and fallback logic here.

Reply:
I think we are ok with a task not being successfully reported indefinitely? As we consider that quite unlikely? 

The retry behaviour here is at least bounded but the result is still a task stuck indefinitely. So I would say retrying indefinitely is not worse.